CSUB-471: Proof of Stake

nathanwhit opened the pull request:
# Description of proposed changes

<describe what this PR is about and why we want it>

---
Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

nathanwhit committed:
Integrate grandpa finality

nathanwhit committed:
Integrate babe consensus

nathanwhit committed:
Integrate staking + dynamic validator selection

nathanwhit committed:
Add some comments on open questions for PoS

nathanwhit committed:
Bump versions

nathanwhit committed:
Revert "Remove PoW code and deps"

This reverts commit f814463bb32d6039131497351e268672cb91ef6f.

nathanwhit committed:
Add pos-switch pallet for runtime support of transition

nathanwhit committed:
Switch to substrate fork

nathanwhit committed:
Runtime migration to init sessions & staking

nathanwhit committed:
Switch from PoW to PoS on the fly

nathanwhit committed:
Remove dead code

nathanwhit committed:
Clean up logging

nathanwhit committed:
Don't spawn initializers as essential, make blocking

nathanwhit committed:
Seamlessly switch authorship when running dev node

nathanwhit committed:
Add chainspec for testing the switch

nathanwhit committed:
Fixup after rebase

nathanwhit committed:
Switch to gluwa fork of substrate

nathanwhit committed:
Fix test client

nathanwhit committed:
Set a 15 second block time

nathanwhit committed:
Use fixed substrate_wasm_builder

and bump rust nightly version

nathanwhit committed:
Add missing try-runtime feature

nathanwhit committed:
New (dev) chains should start as PoS

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

atodorov commented on Cargo.toml:
```diff
@@ -56,74 +57,89 @@ primitives = { path = "primitives", default-features = false }
-sc-cli = { branch = "polkadot-v0.9.40", git = "https://github.com/paritytech/substrate.git" }
-sc-client-api = { branch = "polkadot-v0.9.40", git = "https://github.com/paritytech/substrate.git" }
-sc-client-db = { branch = "polkadot-v0.9.40", git = "https://github.com/paritytech/substrate.git", features = [
+frame-benchmarking-cli = { branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git" }
```
I'd prefer if the `pos-keep-history` branch contains an explicit version name and is created from the relevant polkadot-v* branch. For example:
- `pos-keep-history-polkadot-v0.9.40` is based on upstream `polkadot-v0.9.40` branch
- `pos-keep-history-polkadot-v0.9.41` is based on upstream `polkadot-v0.9.41` branch

and so on. 

This will make versioning explicit and also allow us to "rebase" the `pos-keep-history` branches independently of Creditcoin to facilitate easier dependency upgrades.

atodorov replied:
The current `pos-keep-history` branch seems to be based on `polkadot-v0.9.40` but we have `polkadot-v0.9.41` already available (no changes required to Creditcoin, see #1029).  We'll need at least v0.9.42 before we can drop the nightly compiler.

Do you mind forking off of `polkadot-v0.9.41`  as mentioned above ?

nathanwhit replied:
Ah that's a good idea, that's a lot clearer. I've changed the branch name to include the version, and also updated it to `0.9.41`.

atodorov commented on Cargo.toml:
```diff
@@ -56,74 +57,89 @@ primitives = { path = "primitives", default-features = false }
-sc-client-api = { branch = "polkadot-v0.9.40", git = "https://github.com/paritytech/substrate.git" }
-sc-client-db = { branch = "polkadot-v0.9.40", git = "https://github.com/paritytech/substrate.git", features = [
+frame-benchmarking-cli = { branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git" }
+frame-benchmarking = { branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git", default-features = false }
```
I'd like to enable as many of the upstream CI test jobs against our fork as possible.  So that we can have some sort of confidence when upgrading this branch. 

@nathanwhit can you enable GitHub actions executions on the forked repository ? 

I will take a look at their test jobs in the mean time.

nathanwhit replied:
Hmm according to the repo settings, they should be enabled. I'll take another look, but yeah we'll have to figure out how to run it since parity does CI runs through gitlab instead of GHA.

atodorov replied:
I see some GHA notifications on the branch so CI jobs must be running. I will open another ticket for myself to dig into this further. It wouldn't be as easy as I though initially.

atodorov commented on pallets/pos-switch/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+[package]
```
Should be added to dependabot.yml

atodorov commented on runtime/Cargo.toml:
```diff
@@ -26,7 +26,7 @@ pallet-balances = { workspace = true }
 pallet-difficulty = { workspace = true }
 pallet-offchain-task-scheduler = { workspace = true }
-pallet-insecure-randomness-collective-flip = { workspace = true }
+# pallet-insecure-randomness-collective-flip = { workspace = true }
```
Looks like a leftover comment. If not needed let's remove it. See CSUB-453 as well.

nathanwhit replied:
Yeah, I see it's actually unused now that I'm looking at it. I've removed it now

atodorov commented on pallets/pos-switch/src/lib.rs:
```diff
@@ -0,0 +1,83 @@
+	impl<T: Config> Pallet<T> {
+		/// Switch to PoS
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::BlockWeights::get().max_block, DispatchClass::Operational))]
```
This needs to be converted into WeightInfo / benchmark in order to satisfy the CI scripts.

nathanwhit replied:
Hmm this extrinsic is always going to take up the entire block (just like the `set_code` extrinsic), so there isn't much use in benchmarking it. I pushed a commit to just whitelist the pallet but I can revert and figure something else out if you'd prefer.

atodorov replied:
I'm fine with whitelisting it. Maybe add the explanation above to the git commit message for future reference ?

atodorov commented on ci/env:
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
 
-export RUSTC_VERSION=nightly-2022-12-18
+export RUSTC_VERSION=nightly-2023-04-16
```
Q: is this required b/c of the changes or simply bumping the compiler version for good measure?

nathanwhit replied:
Yeah, CI was failing due to a compiler bug in the version of nightly we were using. I'm not exactly sure why these changes cause it to occur, but after looking into it the easiest solution is to just update the nightly version (since the bug has been fixed since then).

atodorov commented on Cargo.toml:
```diff
@@ -56,74 +57,89 @@ primitives = { path = "primitives", default-features = false }
+tracing = { package = "sp-tracing", branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git", default-features = false }
+frame-try-runtime = { branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git", default-features = false }
+sp-externalities = { branch = "pos-keep-history", git = "https://github.com/gluwa/substrate.git", default-features = false }
+substrate-wasm-builder = "10.0.0"
```
@nathanwhit can you elaborate more on this change. What does it achieve and why is it needed ? 

I'm not sure how this will play out with the upgrade to polkadot-v0.9.42 and switching to a stable rustc version.

nathanwhit replied:
Yeah, so CI was failing due to a compiler bug in the version of nightly we were using. I updated the nightly version to fix that compiler error, but then that hits the bug discussed in https://gluwa.slack.com/archives/C03MQ532BGA/p1681225667436109. The fix for that bug is to update the `substrate-wasm-builder` dependency. They already released a new version, and (unlike most other substrate deps) it's basically standalone so it doesn't need to be on the exact same version as the rest of the substrate deps.

atodorov commented on pallets/pos-switch/src/lib.rs:
```diff
@@ -0,0 +1,83 @@
+				T::RuntimeBlockNumber::from_ref(&block_number).clone().into();
+			SwitchBlockNumber::<T>::put(block_number);
+
+			frame_system::Pallet::<T>::can_set_code(&code)?;
```
Q: Is the `code` argument the WASM runtime or a block hash where a new runtime with consensus switch was written on-chain ?

nathanwhit replied:
The WASM runtime. This extrinsic is basically just `System.set_code` plus some extras.

atodorov replied:
A few more questions here: 

0) I am assuming `switch_to_pos()` will trigger migrations then (by extension of calling `set_code`).  Is that correct?

1) Should we update https://github.com/gluwa/creditcoin/blob/dev/scripts/js/src/runtimeUpgrade.ts#L118 ?

2) Do I understand correctly that we need to call `switch_to_pos()` only once and then it's not needed? 

3) What happens after `switch_to_pos()` is already called as part of scheduling the upgrade to mainnet ? Do we just remove the `pos-switch` pallet ? It sounds like we don't need it afterwards.

nathanwhit replied:
0. Yup
1. Hmmm we'll probably want to do the switch manually since we'll have to upgrade the docker images on our nodes before we can actually do the runtime upgrade on dev.
2. Yup
3. We'd still need the storage item (`SwitchBlockNumber`) but we could remove the `switch_to_pos` extrinsic since it wouldn't be needed.

atodorov replied:
Even more questions:

For 0 and 1) - I think I'll need to modify `runtimeUpgrade.ts` anyway, b/c that's the entry-point for fork-and-migrate tests. Will see how that goes. 

About docker images:

1) Is consensus engine a client side or a runtime component ? I think it is predominantly client side.

2) Assume we upgrade our own nodes with a newer docker image, then call `switch_to_pos()` with a newer runtime. What happens with all of the other nodes which haven't updated to the latest docker image? 

  -- will they still be able to keep in sync with mainnet ? I think yes.
  -- they should definitely not be able to produce blocks b/c the consensus engine has changed. Will these nodes start a fork in the blockchain (a lot of them should be peer connected). Are you aware of any negative consequences that may happen ? 
  -- what will a node operator observe if they aren't aware that consensus has changed  and they haven't upgraded their docker images ?
  -- does the versioning protection from Substrate kick in in this case and how exactly is it going to work? Or will the versioning protection even matter b/c of the consensus change ? 


For 2) and 3) - captured this in CSUB-494.

nathanwhit replied:
1. Yes, primarily client side
2. Nodes that haven't updated their client will not be able to keep in sync, because syncing the blocks requires having the newer consensus to verify them. This is still, in essence, a hard fork. In theory, they could continue to produce PoW blocks and fork off. To try to avoid that, the `switch_to_pos` extrinsic sets the difficulty to the maximum value which should make producing a PoW block close to impossible.

To a node operator, it will just appear like there are no new blocks until they update their client.

There isn't much version protection in this case, as consensus lives outside the runtime and substrate doesn't actually support switching consensus mid-chain. The ability to switch on the fly is what I've added here (and in our substrate fork) and is the "history preservation" feature.

nathanwhit committed:
Remove unused pallet

nathanwhit committed:
Update substrate to polkadot v0.9.41

nathanwhit committed:
Allow the pos switch pallet to not have benchmarks

The `switch_to_pos` extrinsic will always take up
the entire weight of the block (just like the
set_code extrinsic), so there isn't much use in
benchmarking it. Since that is the only
extrinsic in the pallet, just whitelist it in
our benchmark checks.

nathanwhit committed:
Update fork check script to use gluwa substrate

nathanwhit committed:
Appease clippy

nathanwhit committed:
Add dependabot entry for pos switch pallet

nathanwhit committed:
Add missing `try-runtime` feature flags

atodorov commented on node/rpc/src/task.rs:
```diff
@@ -6,7 +6,7 @@ use sc_rpc::DenyUnsafe;
 use sp_runtime::traits;
 use std::sync::Arc;
-use task_scheduler_runtime_api::TaskApi;
+use task_scheduler_runtime_api::TaskApi as RuntimeTaskApi;
```
Commenting here.  2 failing integration tests: 

```
FAIL src/test/metrics.test.ts
  ● Hashrate prometheus metric works

    expect(received).toBeGreaterThan(expected)

    Expected: > 0
    Received:   0

      23 |             expect(value).toBe(0);
      24 |         } else {
    > 25 |             expect(value).toBeGreaterThan(0);
         |                           ^
      26 |         }
      27 |     }
      28 | });
```

and 

```
FAIL src/test/rpc/creditcoin.test.ts
  ● Creditcoin RPC › hashrate() should return a valid hashrate

    expect(received).toBeGreaterThan(expected)

    Expected: > 0
    Received:   0

      69 |             expect(rateObj.hash_count).toBe(0); // eslint-disable-line
      70 |         } else {
    > 71 |             expect(rateObj.rate).toBeGreaterThan(0);
         |                                  ^
      72 |             expect(rateObj.hash_count).toBeGreaterThan(0); // eslint-disable-line
      73 |         }
      74 |         expect(rateObj.elapsed.secs).toBeGreaterThan(0);
```

Both of them related to the hash rate metrics exposed by the node (node/rpc/src/lib.rs).  I don't see anything obvious here though. 

Is the hashrate measurement functionality rendered incorrect by the PoS switch ?

nathanwhit replied:
The reason is that PoS doesn't require any mining, so we don't produce any hashes and the hashrate will always be zero.

To get CI passing, I guess we can: run in PoW for the integration tests, modify/remove these tests, or figure out a way to only run these tests with PoW.

Any thoughts/preferences?

nathanwhit replied:
I went with 
> run in PoW for the integration tests

for now, in the most recent commit.

atodorov replied:
> The reason is that PoS doesn't require any mining, so we don't produce any hashes and the hashrate will always be zero.

In that case the entire metric becomes useless so IMO it should be removed.  We can track this as a separate ticket if you want and just skip the tests so that we don't make this PR too big.

> I went with  run in PoW for the integration tests

Yikes. Please revert this.  We want to be testing with the PoS consensus in order to be close to reality, not artificially instrument the node binary in order to satisfy the test suite.

atodorov replied:
See https://github.com/gluwa/creditcoin/pull/1032/commits/2e0eaefc524aefc8eed6b99328f975451ca18304 for skipping the 2 failing tests. I need to patch one of them b/c asserting that Prometheus metrics work is still important. The hashrate RPC test will be removed.

nathanwhit replied:
Reverted the commit and pushed commit to skip the tests.

atodorov reviewed (changes requested):
Added comment about 2 failing integration tests. They need more investigation.

@nathanwhit do you mind also rebasing while at it ?

atodorov commented on .github/workflows/ci.yml:
```diff
@@ -236,8 +236,8 @@ jobs:
-          echo "RUSTFLAGS=-Zprofile -Ccodegen-units=1 -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Cpanic=abort -Zpanic_abort_tests" >> "$GITHUB_ENV"
-          echo "RUSTDOCFLAGS=-Zprofile -Ccodegen-units=1 -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Cpanic=abort -Zpanic_abort_tests" >> "$GITHUB_ENV"
+          echo "RUSTFLAGS=-Zprofile -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Cpanic=abort -Zpanic_abort_tests" >> "$GITHUB_ENV"
+          echo "RUSTDOCFLAGS=-Zprofile -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Cpanic=abort -Zpanic_abort_tests" >> "$GITHUB_ENV"
```
@nathanwhit was this an attempt to fix the unit test job ? 

I originally though you were adjusting the flags b/c of a new compiler version but that's probably not the case.

I tried by removing all flags and still got the same result. It looks like an out-of-memory issue to me. 

However OOM is also strange b/c the integration-test job also builds creditcoin-node but doesn't OOM.

nathanwhit replied:
Yeah that was just me trying to see if I could get memory usage down, I've reverted it now.

atodorov committed:
Switch CI runner for unit tests to MacOS 12 to avoid OOM

b/c apparently they have 14 GB of RAM vs only 7 GB on Linux:
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

Commit:
Auto-update creditcoin-js type definitions

nathanwhit committed:
Skip hashrate tests

nathanwhit committed:
Move PoS initialization out of runtime upgrade

nathanwhit committed:
Bump version

Commit:
Auto-update creditcoin-js type definitions

atodorov commented on creditcoin-js/src/interfaces/augment-api-consts.ts:
```diff
@@ -20,6 +20,30 @@ export type __AugmentedConst<ApiType extends ApiTypes> = AugmentedConst<ApiType>
 
 declare module '@polkadot/api-base/types/consts' {
     interface AugmentedConsts<ApiType extends ApiTypes> {
+        babe: {
```
FTR I think it would be a good time to bump the version of creditcoin-js b/c we haven't done so in a long time. Can be either inside of this PR or immediately after.

nathanwhit committed:
Don't hardcode grandpa initial set,

properly handle enable grandpa flag, wire up prometheus metrics
with grandpa

nathanwhit committed:
Adjust blockweight

nathanwhit committed:
Make first election successful by flipping init order

nathanwhit committed:
Update lockfile

Commit:
Auto-update creditcoin-js type definitions

nathanwhit committed:
Use old version of srtool for now
